### DISCUSSION TITLE
Broadcast function usage

### DISCUSSION THREAD

**Discussion by aworobel** [lxzan/gws#93] [Q&A] [answered]
Hello,

Thanks for this library.
I try to use Broadcast on the Best Practice example.
As it works per socket connection, how do you implement the []gws.Conn array to broadcast to it and manage the users disconnection ?
I saw the sessions storage section but i didn't find a way to use it properly.

Thanks in advance for your help :)

Arnaud.

**Comment by lxzan** [comment, marked as answer]
```go
package main

import (
	"github.com/lxzan/gws"
)

func GetValueFromSession[T any](s gws.SessionStorage, k string) (value T, ok bool) {
	if v, exist := s.Load(k); exist {
		value, ok = v.(T)
	}
	return
}

type Handler struct {
	gws.BuiltinEventHandler
	conns *gws.ConcurrentMap[int64, *gws.Conn]
}

func (c *Handler) OnOpen(socket *gws.Conn) {
	userId, _ := GetValueFromSession[int64](socket.Session(), "userId")
	c.conns.Store(userId, socket)
}

func (c *Handler) OnClose(socket *gws.Conn, err error) {
	userId, _ := GetValueFromSession[int64](socket.Session(), "userId")
	c.conns.Delete(userId)
}

func (c *Handler) Broadcast(p []byte) {
	var b = gws.NewBroadcaster(gws.OpcodeText, p)
	defer b.Close()
	c.conns.Range(func(userId int64, socket *gws.Conn) bool {
		_ = b.Broadcast(socket)
		return true
	})
}
```

**Comment by aworobel** [comment]
Hello,

Thank you a lot for your valued assistance, i will check now how to pass an userId from the client webpage with socket.io and get it in the badckend.
For the moment i get a null pointer exception (normal, userId should be set/get before) and tried to deal with [this example](https://github.com/lxzan/gws/blob/main/examples/chatroom/main.go).
